### PR TITLE
Collapsible Side Panel

### PR DESCRIPTION
**Pull request by apiraam96**
Screenshot 2024-02-21 at 12 56 05 PM

Closes Issue #86

### CODE REVIEW

File: Eplant/UI/Sidebar.tsx [hunk 1]
```diff
@@ -1,48 +1,70 @@
...
 import { LeftNav } from './LeftNav'
 import ResponsiveDrawer from './ResponsiveDrawer'
 
+
+// This sidebar width is hard coded --> not responsive to screen size
 export const sidebarWidth = 300
 
-export default function Sidebar() {
+
+export default function SideBar() {
```
Comment (zkdan) [inline Eplant/UI/Sidebar.tsx]: Use Sidebar instead of SideBar to remain consistent.

File: Eplant/UI/LeftNav/index.tsx [hunk 2]
```diff
@@ -29,6 +35,16 @@ export function LeftNav(props: {
   const [geneticElements, setGeneticElements] = useGeneticElements()
   const [darkMode, setDarkMode] = useDarkMode()
 
+  //Grabbing collapse state and setter function from useContext hook
+  const context = useContext(collapseContext)
```
Comment (zkdan) [inline Eplant/UI/LeftNav/index.tsx]: What's the logic behind using useContext instead of Jotai?

File: Eplant/UI/Sidebar.tsx (hunk 1 above)
Comment (zkdan) [inline Eplant/UI/Sidebar.tsx]: Revert this back to Sidebar.

In general, don't change component names unless you absolutely must and in that case, change the file name to match.

File: Eplant/Eplant.tsx [hunk 3]
```diff
@@ -30,11 +32,28 @@ export default function Eplant() {
  * The main Eplant component. This is the root of the application. It contains the left nav and the layout.
  * @returns {JSX.Element} The rendered Eplant component
  */
+
+// createContext type
+export type CollapseContent = {
```
Comment (zkdan) [inline Eplant/Eplant.tsx]: Use a Jotai atom instead

File: Eplant/UI/LeftNav/index.tsx [hunk 4]
```diff
@@ -38,8 +54,19 @@ export function LeftNav(props: {
   }, [geneticElements])
   return (
     <Stack gap={2} direction='column' height={'100%'}>
-      <LogoWithText text='ePlant' />
-      <SearchGroup
+      <div style={{ display: 'flex', justifyContent: 'space-between' }}>
+        {!collapse ? <LogoWithText text='ePlant' /> : <LogoWithText text='' />}
+        <button style={
+          !collapse ? { backgroundColor: 'transparent', border: 'none', transform: 'translateX(+20%)' } : { backgroundColor: 'transparent', border: 'none', transform: 'translateX(-30%)', transition: 'all 1s ease-out' }
```
Comment (zkdan) [inline Eplant/UI/LeftNav/index.tsx]: Is it possible to have the ternary start with the positive case instead of the negative one? So,  something like`collapse ?  { backgroundColor: 'transparent', border: 'none', transform: 'translateX(-30%)', transition: 'all 1s ease-out' }: { backgroundColor: 'transparent', border: 'none', transform: 'translateX(+20%)' }`?

Comment (zkdan) [inline Eplant/UI/LeftNav/index.tsx]: `translateX` doesn't require the `+` for a positive value

Comment (zkdan) [reply Eplant/UI/LeftNav/index.tsx]: Or alternatively, have a local variable called something like `sidebarOpen` whose value is `!collapse`

Comment (zkdan) [inline Eplant/UI/LeftNav/index.tsx]: Usually the `transition` rule is added to the initial state of the element - is that the case here?

File: Eplant/UI/LeftNav/index.tsx [hunk 5]
```diff
@@ -60,12 +87,18 @@ export function LeftNav(props: {
           }
         }}
       />
-      <Collections
-        onSelectGene={props.onSelectGene}
-        selectedGene={props.selectedGene}
-      />
+        <Collections
+          onSelectGene={props.onSelectGene}
```
Comment (zkdan) [inline Eplant/UI/LeftNav/index.tsx]: Destructure at the top of the file to avoid having any variables called `props.whatever`

File: Eplant/UI/LeftNav/index.tsx [hunk 6]
```diff
@@ -60,12 +87,18 @@ export function LeftNav(props: {
...
+          selectedGene={props.selectedGene}
+        /> </>}
+
+      {collapse && <>
+        <div onClick={() => setCollapse(false)} style={{ marginLeft: '10px' }}>
+          <SearchIcon sx={{ '&:hover': { cursor: 'pointer', transform: 'scale(1.1)' } }} />
+        </div>
+      </>}
```
Comment (zkdan) [inline Eplant/UI/LeftNav/index.tsx]: The fragment is redundant if you're only return the one `div`

File: package.json [hunk 7]
```diff
@@ -73,4 +74,4 @@
     "vite-plugin-eslint": "^1.8.1",
     "vite-tsconfig-paths": "^4.2.1"
   }
-}
```
Comment (zkdan) [inline package.json]: Revert this so as to not upload unnecessary changes (the`package-lock.json` file should remove itself at the same time)

File: Eplant/UI/LeftNav/index.tsx [hunk 8]
```diff
@@ -75,7 +108,10 @@ export function LeftNav(props: {
           }
           label='Dark mode'
         />
-      </FormGroup>
+      </FormGroup> */}
+      <div style={{ display: 'flex', justifyContent: 'center' }} onClick={() => setDarkMode(!darkMode)}>
+        <DarkModeIcon sx={darkMode ? { '&:hover': { cursor: 'pointer', color: 'white', transform: 'scale(1.1)' } } : { '&:hover': { cursor: 'pointer', color: 'black', transform: 'scale(1.1)' } }} color={darkMode ? 'primary' : 'secondary'} />
```
Comment (zkdan) [inline Eplant/UI/LeftNav/index.tsx]: Ensure that the hover styles apply when the element is focused, as well

File: Eplant/UI/LeftNav/index.tsx [hunk 9]
```diff
@@ -60,12 +87,18 @@ export function LeftNav(props: {
...
+        /> </>}
+
+      {collapse && <>
+        <div onClick={() => setCollapse(false)} style={{ marginLeft: '10px' }}>
+          <SearchIcon sx={{ '&:hover': { cursor: 'pointer', transform: 'scale(1.1)' } }} />
+        </div>
+      </>}
       <Box flexGrow={1} />
-      <FormGroup>
+      {/* <FormGroup>
```
Comment (zkdan) [inline Eplant/UI/LeftNav/index.tsx]: Check with Nick on whether we're removing the toggle completely. If so, delete it rather than commenting it out.

File: Eplant/UI/Sidebar.tsx [hunk 10]
```diff
@@ -1,48 +1,70 @@
+import { useState } from 'react'
+import { useContext } from 'react'
+import { Sidebar } from 'react-pro-sidebar'
```
Comment (zkdan) [inline Eplant/UI/Sidebar.tsx]: Where is this Sidebar being used?

File: Eplant/UI/Sidebar.tsx [hunk 11]
```diff
@@ -1,48 +1,70 @@
...
+import { Menu } from 'react-pro-sidebar'
```
Comment (zkdan) [inline Eplant/UI/Sidebar.tsx]: Where is this Menu being used? Does `react-pro-sidebar` do something we can't do with MUI?

**Review by zkdan — CHANGES_REQUESTED** [review]
A good start! The moon icon is a nice touch. 

I believe the takeaways from the meeting on 2/22/24 were:
- The sidebar content being invisible when the sidebar is collapsed
- The icon to slide the sidebar in and out should be larger and visually in its own container like a tab on a file
- The logo should always be visible

File: Eplant/UI/Sidebar.tsx (hunk 11 above)
Comment (apiraam96) [reply Eplant/UI/Sidebar.tsx]: I removed the package now (this was just a package I was playing around with)

File: Eplant/UI/LeftNav/index.tsx (hunk 2 above)
Comment (apiraam96) [reply Eplant/UI/LeftNav/index.tsx]: Refactored code to use Jotai

File: Eplant/UI/LeftNav/index.tsx (hunk 4 above)
Comment (apiraam96) [reply Eplant/UI/LeftNav/index.tsx]: Yup, it's a turnery operator for 2 different style components

File: Eplant/EplantLayout.tsx [hunk 12]
```diff
@@ -50,6 +66,7 @@ const EplantLayout = () => {
           justifyContent: 'stretch',
         }}
       >
+        <div style={{ color: 'red' }} />
```
Comment (zkdan) [inline Eplant/EplantLayout.tsx]: What's this `color:red` for?

File: Eplant/UI/LeftNav/index.tsx [hunk 13]
```diff
@@ -36,40 +59,109 @@ export function LeftNav(props: {
...
-      />
-      <Box flexGrow={1} />
-      <FormGroup>
-        <FormControlLabel
-          control={
-            <Switch
-              onChange={(v) => setDarkMode(v.target.checked)}
-              checked={darkMode}
+          onClick={settingCollapse}
+          onMouseEnter={() => setTabWidth('40px')}
```
Comment (zkdan) [inline Eplant/UI/LeftNav/index.tsx:83]: This looks great on the UI!

**Review by zkdan — CHANGES_REQUESTED** [review]
You're not using the `react-pro-sidebar`, right? I bet you just need to delete your `package-lock` file so that one can be regenerated that will have no conflicts with staging. If you're having trouble, you may want to totally delete `node_modules` and regenerate everything from staging's `package` file.

It looks great though!

File: Eplant/EplantLayout.tsx (hunk 12 above)
Comment (apiraam96) [reply Eplant/EplantLayout.tsx]: Sorry, this was just a styling test

File: Eplant/UI/LeftNav/index.tsx (hunk 13 above)
Comment (apiraam96) [reply Eplant/UI/LeftNav/index.tsx:83]: Thank you! I removed the package-lock.json and regenerated it via npm install

File: Eplant/Eplant.tsx [hunk 14]
```diff
@@ -28,6 +30,14 @@ export default function Eplant() {
  * The main Eplant component. This is the root of the application. It contains the left nav and the layout.
  * @returns {JSX.Element} The rendered Eplant component
  */
+
+// Setting the sidebar collapse state with jotai
+export const isCollapse = atom(false)
```
Comment (Yukthiw) [inline Eplant/Eplant.tsx]: Can you put this in `Eplant/state/index.tsx`. You also don't need to define a separate setter atom. You should just define a custom hook called useSidebarState (or something along those lines) and use this whenever the collapse state needs to be read or written to. 

See [the use darkMode hook](https://github.com/BioAnalyticResource/ePlant/blob/98fc1a21b622a2bad6af33f95d19810e87142c22/Eplant/state/index.tsx#L210) for how to create a custom hook (just use a regular atom like you have instead of "atomWithOptionalStorage") and see [the darkMode toggle](https://github.com/BioAnalyticResource/ePlant/blob/98fc1a21b622a2bad6af33f95d19810e87142c22/Eplant/UI/LeftNav/index.tsx#L29) for how to use it.

In the above example a separate setter atom is defined but it is unnecessary for the sidebar state.

File: Eplant/EplantLayout.tsx [hunk 15]
```diff
@@ -29,15 +40,19 @@ const EplantLayout = () => {
...
+  //Initializing isCollapse variable to isCollapse jotai atom state
+  const [isCollapse] = useAtom(isCollapseState)
+
   return (
     <Box
       sx={(theme) => ({
-        height: `100%`,
-        left: `${sidebarWidth}px`,
+        height: `calc(100% - ${theme.spacing(1)})`,
+        left: `${isCollapse ? 100 : sidebarWidth}px`,
```
Comment (Yukthiw) [inline Eplant/EplantLayout.tsx]: Nit: Can you define another variable called collapsedSidebarWidth and put it in `Eplant/UI/Sidebar.tsx`

**Review by Yukthiw — CHANGES_REQUESTED** [review]
Looks great, just a note on how to use Jotai consistent with the rest of the project.

**Review by Yukthiw — APPROVED** [review]
LGTM

**Review by zkdan — APPROVED** [review]

**Comment by apiraam96** [comment]
@asherpasha

**Comment by asherpasha** [comment]
For some reason GitHub Actions checks are not running for this pull request. Two ways to merge this:

1) I disable branch protection, merge this pull request and re enable branch protection.
Or,
2). Back up your work. Close this pull request. Open a new pull request. And then check should runs (hopefully).

Your call.

**Comment by asherpasha** [comment]
I think workflows need to be updated to run GitHub checks on pull request. Currently, it's only on push.

See BAR API example:
https://github.com/BioAnalyticResource/BAR_API/blob/dev/.github/workflows/bar-api.yml

**Comment by Yukthiw** [comment]
> I think workflows need to be updated to run GitHub checks on pull request. Currently, it's only on push.
> 
> See BAR API example: https://github.com/BioAnalyticResource/BAR_API/blob/dev/.github/workflows/bar-api.yml

I think there's a setting change that needs to be done, having a PR trigger just runs the action when you first open a PR which is not what we want.  [See this StackOverflow post](https://stackoverflow.com/questions/73314842/unable-to-trigger-github-actions-when-pull-request-comes-from-a-forked-repositor) @asherpasha

**Comment by asherpasha** [comment]
This is done. Please try again.

**Comment by apiraam96** [comment]
I'm still having the pending problem with the build and lint-and-format checks. It looks like there are other PRs from Jamie and Mike that want to merge their forked branches to staging too
